lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -0,0 +1,144 @@
+-- mkNetworkLayer _httpBridge = NetworkLayer
+--     { nextBlocks = undefined
+--     , networkTip = undefined
+--     , postTx = undefined
+--     }
+--
+-- -- | Creates a jormungandr 'NetworkLayer' using the given connection
+-- -- settings.
+-- newNetworkLayer
+--     :: Int -> IO (NetworkLayer IO)
+-- newNetworkLayer port = mkNetworkLayer <$> newHttpBridge port
+--   where
+--     newHttpBridge = undefined
+
+{-------------------------------------------------------------------------------
+                            Jormungandr Client
+-------------------------------------------------------------------------------}
+
+-- | Endpoints of the jormungandr REST API.
+newtype Jormungandr m = Jormungandr
+    { getTipId
+        :: ExceptT ErrNetworkUnreachable m BlockId
+    }
+
+-- | Construct a 'Jormungandr'-client
+--
+-- >>> mgr <- newManager defaultManagerSettings
+-- >>> j = mkJormungandr mgr (BaseUrl Http "localhost" 8080 "")
+-- >>> runExceptT $ getTipId j
+-- Right (BlockId (Hash {getHash = "26c640a3de09b74398c14ca0a137ec78"}))
+mkJormungandr
+    :: Manager -> BaseUrl -> Jormungandr IO
+mkJormungandr mgr baseUrl = Jormungandr
+    { getTipId = ExceptT $ do
+        let ctx = safeLink api (Proxy @GetTipId)
+        run cGetTipId >>= defaultHandler ctx
+    }
+  where
+    run :: ClientM a -> IO (Either ServantError a)
+    run query = retryEither $ runClientM query (mkClientEnv mgr baseUrl)
+
+
+    defaultHandler
+        :: Link
+        -> Either ServantError a
+        -> IO (Either ErrNetworkUnreachable a)
+    defaultHandler ctx = \case
+        Right c -> return $ Right c
+
+        -- The node has not started yet or has exited.
+        -- This could be recovered from by either waiting for the node
+        -- initialise, or restarting the node.
+        Left (ConnectionError e) ->
+            return $ Left $ ErrNetworkUnreachable e
+
+        -- Other errors (status code, decode failure, invalid content type
+        -- headers). These are considered to be programming errors, so crash.
+        Left e -> do
+            throwM (ErrUnexpectedNetworkFailure ctx e)
+
+    cGetTipId = client api
+
+data ErrUnexpectedNetworkFailure
+    = ErrUnexpectedNetworkFailure Link ServantError
+    deriving (Show)
+
+instance Exception ErrUnexpectedNetworkFailure
+
+retryEither :: IO (Either ServantError a) -> IO (Either ServantError a)
+retryEither action =
+    retrying
+        (exponentialBackoff 1000000 <> limitRetries 4)
+        shouldRetry
+        (const action)
+  where
+    shouldRetry _ (Right _) = return False
+    shouldRetry status (Left err) = do
+        let cumDelay = fromIntegral @_ @Double (rsCumulativeDelay status) / 1000000
+        let n = rsIterNumber status
+        let info = if n > 0
+            then "(retry "+||n||+ " after "+||cumDelay||+ "s)"
+            else mempty
+        TIO.putStrLn $ "[ERROR] Jormungandr"<>info<>": "+||err||+""
+
+        return True
```

Comment:
What about: 

- retrying only on `ConnectionFailure`, but not for all `Left`
- using combinators provided by the `retry` library like [limitRetriesByCumulativeDelay](https://hackage.haskell.org/package/retry-0.8.0.1/docs/Control-Retry.html#v:limitRetriesByCumulativeDelay) ?

Reply:
I am thinking something like this
```haskell
    { getTipId = do
        let ctx = safeLink api (Proxy @GetTipId)
        run cGetTipId ctx $ \case
            Right a -> Proceed $ return a
            Left (ConnectionError _) -> Retry
            Left e -> Proceed $ throwM ErrUnexpectedNetworkFailure ctx e
```

`limitRetriesByCumulativeDelay` might make sense indeed`. Maybe [capDelay](https://hackage.haskell.org/package/retry-0.8.0.1/docs/Control-Retry.html#v:capDelay) too